Persist Pivot Table Column Widths

iethree opened the pull request:
resolves https://github.com/metabase/metabase/issues/27598

## Description

Adds a new `pivot_table.column_widths` visualization setting and uses it to persist column widths.

![persist resize](https://user-images.githubusercontent.com/30528226/213761518-f7f782c7-eded-425c-b67f-f5f3bc8ce17c.gif)


## Testing Steps
- open a gnarly pivot table with multiple aggregations
- resize some columns
- refresh the page, see that the settings persist via query serialization in the url
- save the query
- refresh the page again and see that the settings persist in the db

iethree committed:
add PivotTable unit tests

iethree committed:
measure leftHeader cell content

iethree committed:
test cell data width detection

iethree committed:
obey the linter

iethree committed:
convert PivotTable to typescript

iethree committed:
remove unused hasCustomColors

iethree committed:
add drag resizing

iethree committed:
address review comments

iethree committed:
User Resizable Pivot Table Columns (#27692)

* move column widths to state

* add drag resizing

* only get initial header widths on row change

* better column change handling

* return if there are no valid rowIndexes

* address review comments

iethree committed:
persist column width settings

alxnddr reviewed (commented):
@iethree can we store only changed column widths? It would help to avoid populating viz settings with duplicated default values:

```{"dataset_query":{"database":1,"type":"query","query":{"source-table":2,"aggregation":[["count"],["avg",["field",10,null]],["avg",["field",10,null]]],"breakout":[["field",17,{"binning":{"strategy":"default"}}],["field",1,{"source-field":9}],["field",14,{"temporal-unit":"month"}]]}},"display":"pivot","displayIsLocked":true,"parameters":[],"visualization_settings":{"pivot_table.column_split":{"rows":[["field",17,{"binning":{"strategy":"num-bins","min-value":0,"max-value":70,"num-bins":8,"bin-width":10}}]],"columns":[["field",14,{"temporal-unit":"month"}],["field",1,{"source-field":9}]],"values":[["aggregation",0],["aggregation",1],["aggregation",2]]},"pivot_table.column_widths":{"leftHeaderWidths":[107],"totalLeftHeaderWidths":107,"valueHeaderWidths":[100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,240,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100,100]},"table.pivot_column":"CATEGORY","table.cell_column":"count"},"original_card_id":33}```

iethree committed:
update unit tests

iethree commented:
> @iethree can we store only changed column widths? It would help to avoid populating viz settings with duplicated default values:


👍  handled this in the base branch 836e91c7f85ffd6742f00079b12c89bcf21c1085

iethree committed:
fix rebase

iethree committed:
e2e test column resize and persistence

iethree committed:
keep column width state locally unless changed by the user

this handles a couple situations:
 (1) where an existing pivot table is opened and is magically "changed" because we auto-measured header widths and loaded that data into visualization_settings, and
 (2) where, like pinned cards in dashboards, there is no updatable state for visualization settings, we need to be able to do measurements and update it within the component and can't rely on mutable state of a parent component

iethree commented:
re-requesting review here as the implementation changed a bit with https://github.com/metabase/metabase/pull/27800/commits/92b75b9d704e2194c7209c46bfbbfea8dd8028d8

I brought column width state back into the component to deal with some (good) failing tests.
